UTRannotator.pm

```diff
@@ -0,0 +1,1351 @@
+        existing_OutOfFrame_oORFs => 'The number of existing out-of-frame overlapping ORFs (OutOfFrame oORF) at the 5 prime UTR',
+        existing_InFrame_oORFs => 'The number of existing inFrame overlapping ORFs (inFrame oORF) at the 5 prime UTR',
+  };
+  return $self->{_header_info};
+}
+
+sub run {
+  my ($self, $tva) = @_;
+
+  #only annotate the effect if the variant is 5_prime_UTR_variant
+  return {} unless grep {$_->SO_term eq '5_prime_UTR_variant'}  @{$tva->get_all_OverlapConsequences};
+
+  my $bvfo = $tva->base_variation_feature_overlap;
+  my $bvfoa = $tva;
+
+  #retrieve the variant info
+  my $vf = $tva->variation_feature;
+  my $chr = ($vf->{chr}||$vf->seq_region_name);
+  my $pos = ($vf->{start}||$vf->seq_region_start);
+  my $ref = $bvfo->get_reference_VariationFeatureOverlapAllele->variation_feature_seq;
+  my $alt = $tva->variation_feature_seq;
+  my %variant = (
+        "chr" => $chr,
+        "pos" => $pos,
+        "ref" => $ref,
+        "alt" => $alt,
+  );
+
+  #retrieve the UTR info: transcript id, strand, five prime UTR sequence, start and end genomic coordinates.
+  my $t = $bvfo->transcript;
+  my $transcript_id = (defined $t? $t->stable_id: undef);
+
+  #retrieve the gene symbol of the transcript
+  my $symbol = $t->{_gene_symbol} || $t->{_gene_hgnc};
+
+  #retrieve the strand of the transcript
+  my $tr_strand = $t->strand + 0;
+  my $cds = $t->translateable_seq();
+
+  #retrieve the five prime utr sequence
+  my $five_prime_seq = (defined $t->five_prime_utr? $t->five_prime_utr->seq(): undef);
+
+  #Type: five_prime_feature - Bio::EnsEMBL::Feature
+  my $UTRs = $t->get_all_five_prime_UTRs();
+  my @five_utr_starts;
+  my @five_utr_ends;
+  foreach  my $utr (@$UTRs){
+    my $start = $utr->start(); # this will return the absolute starting positions in chromosome of the UTR exons
+    my $end = $utr->end(); # this will return the absolute ending positions in chromosome of the UTR exons
+    push(@five_utr_starts, $start);
+    push(@five_utr_ends,$end);
+  }
+
+  my @sorted_starts = sort {$a <=> $b} @five_utr_starts;
+  my @sorted_ends = sort {$a <=> $b} @five_utr_ends;
+
+  my %UTR_info = (
+        "gene" => $symbol,
+        "start" => \@sorted_starts,
+        "end" => \@sorted_ends,
+        "seq" => $five_prime_seq,   #the exon sequences in 5'UTR
+        "strand" => $tr_strand,
+        "cds_seq" => $cds,
+  );
+
+  my %kozak_strength;
+  $kozak_strength{1}='Weak';
+  $kozak_strength{2}='Moderate';
+  $kozak_strength{3}='Strong';
+
+  my %uAUG_gained = %{$self->uAUG_gained(\%variant,\%UTR_info, %kozak_strength)};
+  my %uSTOP_lost = %{$self->uSTOP_lost(\%variant,\%UTR_info, %kozak_strength)};
+  my %uAUG_lost = %{$self->uAUG_lost(\%variant,\%UTR_info, %kozak_strength)};
+  my %uSTOP_gained = %{$self->uSTOP_gained(\%variant,\%UTR_info, %kozak_strength)};
+  my %uFrameshift = %{$self->uFrameshift(\%variant,\%UTR_info, %kozak_strength)};
```

Comment:
These methods always call `get_allele_exon_pos()` and `return {}` if the result is not defined. 
You could do that before line 179 and only call the methods if defined $mut_pos && $end_pos. 
Something like:
  ```
  my %uAUG_gained = {};
  my %uSTOP_lost = {};
  my %uAUG_lost = {};
  my %uSTOP_gained = {};
  my %uFrameshift = {};
  my ($mut_pos, $end_pos) = $self->get_allele_exon_pos($tr_strand, $pos, $ref_coding, \%UTR_info);
  if(defined($mut_pos) && defined($end_pos)) {
    my @sequence = split //, $five_prime_seq;
    my $mut_utr_seq = $self->mut_utr_sequence(\@sequence,$mut_pos,$ref_coding,$alt_coding,$tr_strand);
    %uAUG_gained = %{$self->uAUG_gained(\%variant,\%UTR_info, %kozak_strength, $ref_coding, $alt_coding, $mut_pos, $mut_utr_seq)};
    %uSTOP_lost = %{$self->uSTOP_lost(\%variant,\%UTR_info, %kozak_strength, $ref_coding, $alt_coding, $mut_pos, $mut_utr_seq)};
    %uAUG_lost = %{$self->uAUG_lost(\%variant,\%UTR_info, %kozak_strength, $ref_coding, $alt_coding, $mut_pos, $mut_utr_seq)};
    %uSTOP_gained = %{$self->uSTOP_gained(\%variant,\%UTR_info, %kozak_strength, $ref_coding, $alt_coding, $mut_pos, $end_pos, $mut_utr_seq)};
    %uFrameshift = %{$self->uFrameshift(\%variant,\%UTR_info, %kozak_strength, $ref_coding, $alt_coding, $mut_pos, $end_pos, $mut_utr_seq)};
  }
```